src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs

```diff
@@ -547,80 +549,118 @@ private static void UpdateOptionsForNextIteration(ref ChatOptions options, strin
                 // and then await all of them. We avoid forcibly introducing parallelism via Task.Run,
                 // but if a function invocation completes asynchronously, its processing can overlap
                 // with the processing of other the other invocation invocations.
-                results = await Task.WhenAll(
+                results.AddRange(await Task.WhenAll(
                     from i in Enumerable.Range(0, functionCallContents.Count)
                     select ProcessFunctionCallAsync(
                         messages, options, functionCallContents,
-                        iteration, i, captureExceptions: true, cancellationToken));
+                        iteration, i, captureExceptions: true, isStreaming, cancellationToken)));
+
+                shouldTerminate = results.Any(r => r.ShouldTerminate);
             }
             else
             {
                 // Invoke each function serially.
-                results = new FunctionInvocationResult[functionCallContents.Count];
-                for (int i = 0; i < results.Length; i++)
+                for (int i = 0; i < functionCallContents.Count; i++)
                 {
-                    results[i] = await ProcessFunctionCallAsync(
+                    var functionResult = await ProcessFunctionCallAsync(
                         messages, options, functionCallContents,
-                        iteration, i, captureCurrentIterationExceptions, cancellationToken);
+                        iteration, i, captureCurrentIterationExceptions, isStreaming, cancellationToken);
+
+                    results.Add(functionResult);
+
+                    // If any function requested termination, we should stop right away.
+                    if (functionResult.ShouldTerminate)
+                    {
+                        shouldTerminate = true;
+                        break;
+                    }
                 }
             }
 
-            var shouldTerminate = false;
+            IList<ChatMessage> addedMessages = CreateResponseMessages(results.ToArray());
+            ThrowIfNoFunctionResultsAdded(addedMessages);
+            UpdateConsecutiveErrorCountOrThrow(addedMessages, ref consecutiveErrorCount);
 
-            IList<ChatMessage> added = CreateResponseMessages(results);
-            ThrowIfNoFunctionResultsAdded(added);
-            UpdateConsecutiveErrorCountOrThrow(added, ref consecutiveErrorCount);
-
-            messages.AddRange(added);
-            foreach (FunctionInvocationResult fir in results)
+            if (messages is List<ChatMessage> messagesList)
+            {
+                messagesList.AddRange(addedMessages);
+            }
+            else
             {
-                shouldTerminate = shouldTerminate || fir.ShouldTerminate;
+                foreach (var addedMessage in addedMessages)
+                {
+                    messages.Add(addedMessage);
+                }
             }
```

Comment:
If we don't already have one somewhere that does this (I think we might), we should move this into a helper and dedup the occurrences so that its just `messagesList.AddRange(addedMessages)`, with that AddRange extension method doing the type test and the if/else.

Reply:
I thought about this, but didn't want to introduce more code for such a unit need, I'm happy adding an extension method though.

Reply:
It'll be less code, as you have multiple copies of this pattern now.

It doesn't need to be public, nor does it even need to be an extension... it could just be a private helper.